### PR TITLE
Introduce Introspection covenants

### PR DESCRIPTION
**Pull request by sanket1729**
This PR has refactors and cleanups for introducing new covenants. 
Also adds a fragment for CheckSigFromStack and integration tests testing the new fragment along with its satisfaction.

### CODE REVIEW

File: src/extensions/csfs.rs [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+use crate::miniscript::types::{Base, Correctness, Dissat, ExtData, Input, Malleability};
+use crate::policy::{self, Liftable};
+use crate::util::MsKeyBuilder;
+use crate::{
+    expression, interpreter, miniscript, Error, Extension, ForEach, MiniscriptKey, Satisfier, Tap,
+    ToPublicKey, TranslatePk, Translator,
+};
+
+/// Version struct
+/// `DEPTH <12> SUB PICK <num> EQUAL`
```
Comment (apoelstra) [inline src/extensions/csfs.rs:25]: In 761799269e0b5bf29a7a9fa6162a468653a5f351:

Comment also out of date.

File: src/extensions/csfs.rs [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Clone, Copy)]
+pub struct CheckSigFromStack<Pk: MiniscriptKey> {
+    pub pk: Pk,
```
Comment (apoelstra) [inline src/extensions/csfs.rs:28]: In 761799269e0b5bf29a7a9fa6162a468653a5f351:

Should this fragment also enforce the signed message? In which case we'd need another field in this struct and updates to all the script serialization etc..

File: src/extensions/csfs.rs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+    fn script_size(&self) -> usize {
+        1 + 32 + 1 // opcode + key+ push
+    }
+
+    fn segwit_ctx_checks(&self) -> Result<(), miniscript::context::ScriptContextError> {
+        // New opcodes only supported in taproot context
+        Err(ScriptContextError::ExtensionError(
+            "CSFS only available in Taproot".to_string(),
+        ))
```
Comment (apoelstra) [inline src/extensions/csfs.rs:133]: In 761799269e0b5bf29a7a9fa6162a468653a5f351:

In the next commit you add this to `AllCov` and (as I understand it) the legacy covenant structure ... which is segwit only. But here you're saying it's taproot only? I'm a bit confused.

If this is a segwit thing we should remember that there is no such thing as a non-verify CSFS.

File: src/extensions/csfs.rs [hunk 4]
```diff
@@ -0,0 +1,266 @@
+//! Miniscript extension: ver_eq
+//! Note that this fragment is only supported for Segwit context
+//! You are most likely looking for taproot direct tx introspection
```
[the remark this answers is not in the document]
Comment (sanket1729) [reply src/extensions/csfs.rs:3]: Good catch. This is a copy paste error

File: src/extensions/csfs.rs (hunk 3 above)
Comment (sanket1729) [reply src/extensions/csfs.rs:133]: We can add all fragments to `CovenantExt`. But the type system would not allow creation of these. 
- Miniscript<Pk, SegwitV0, CovenantExt<Pk>> when parsing csfs fragment. Similar to how to `multi_a` is disallowed. There is a test also checking this.
- `Miniscript<Pk, Tap, VerEq>` as that only available in Segwtiv0. 


Perhaps, we should rename VerEq to LegacyVerEq to clarify that this is based on old CSFS design paradigm.

Comment (apoelstra) [reply src/extensions/csfs.rs:133]: Yes, I'd like everything based on the old CSFS paradigm to have `Legacy` in the name.

And thanks for the explanation of how the type system keeps everything in bounds here. Very cool.

**Comment by sanket1729** [comment]
@apoelstra. Sorry for the delay and the long diff. I finally managed a way that works. The last commit has some examples. Overall, I think this cleans up a lot of weird code because we were mixing `MinisciptKey` and `Extension`. Cleanly separating them really simplifies things as seen in the last test case.

`Extension` is the trait denoting which extensions are enabled. `ExtParam` is a trait for arguments to extensions.

**Review by apoelstra — APPROVED** [review]
ACK a72ff21bcd13329d4e63b4df374f9a42f2d8324f

**Comment by apoelstra** [comment]
Nice! This is a very long diff. I'm not entirely satisfied with the satisfier API having extra stuff for every fragment but maybe that's the right design choice.

**Pull request merged**